README.md
# MCASM—A Microcode Assembler for Hobbyist Projects

This is a simple microcode assembler (or microassembler). It can be used to
generate wide, horizontal microcode ROMs for relatively simple CPU
architectures. It's mainly intended for hobbyists working on home-designed
CPUs. Using a description of the microcode, it generates one or more ROM,
EPROM, EEPROM, Flash et cetera images suitable for inclusion in a software
emulator, a hardware device, or for use in a hardware description language such
as Verilog or VHDL.

For a full description, please visit https://www.bedroomlan.org/projects/mcasm/.
 
This tool was written to build the microcode for my scratch-built 16-bit
minicomputer, the CFT. Read about that at https://www.bedroomlan.org/cft/.

## Requirements

To run mcasm, you will need:

* Python 3.
* The C Preprocessor (`cpp`). This usually comes with your C compiler.

## Microcode in Theory

Microcode is a collection of *microprograms*. Each microprogram depends on a
number of *conditions* and is made up of a number of *microinstructions*. Each
microinstruction is in turn made up of a number of *signal* specifications.

*Microprograms* implement basic functions. In most cases, each microprogram
will implement one machine instruction. This, of course, depends on the machine
architecture.

*Conditions* are inputs to the computer's sequencer. They include major states
of the machine such as whether an interrupt is being serviced, whether the
machine is being reset, et cetera. A microprogram counter (the offset into the
current microprogram) is also part of the conditions. Conditions may be
don't-care values. Conditions are grouped in multi-bit fields for convenience
or to define multi-bit values like opcodes. Microprograms are selected and
executed when the conditions match. The matching itself is down to the
hardware. For instance, to match an opcode field, a value from the opcode field
of the instruction register would be fed to the corresponding address lines in
a microcode ROM.

*Microinstructions* perform tasks simultaneously by directly manipulating the
sequencer by (you guessed it), signals. Signals may be asserted or
de-asserted. Some signals are active-high, others are active-low. Signals may
be grouped together into *fields* if they are semantically similar,
mutually-exclusive, or if they represent a multi-bit field such as a register
number.

This, of course, is theory. The assembler may generate microcode that is used
in vastly different ways. The architecture determines how the bits produced by
`mcasm` are interpreted in practice.

## Preprocessor

A lot of microcode writing can be repetitive, and repeating structures by hand
makes things prone to errors. Errors in microcode have potentially disastrous
effects, so `mcasm` uses `cpp`, the C Preprocessor to prepare input for
it. That makes the assembler syntax fairly C-like in some respects. As a bonus,
you get parametric macros, conditional compilation, and all the other goodies
`cpp` has to offer.

The syntax of `cpp` is beyond the scope of this article. Please refer to its
documentation.

## Syntax

Here is a very brief example of a microcode input file:

```
// Minimum Microcode file.

cond OP:4;
cond uaddr:3;

signal /MEMIO    = ....1;
signal MREAD     = ...1.;
signal MWRITE    = ..1..;

field  REGMOVE   = XX___;
signal MAR <- PC = 01...;
signal IR <- MEM = 10...;

start OP=XXXX; // A tiny fetch instruction
  /MEMIO, MREAD, MAR <- PC;   // 
  hold;                       // Same as MEMIO, MREAD, MAR <- PC
  hold, -MREAD, IR <- MEM;    // Same as MEMIO, MAR <- PC, IR <- MEM

// End of file.
```

C multi-line comments `/* ... */` and C++ single-line comments `// ...` are
both understood.

All declarations and microinstructions must end in a semicolon.

This example defines two condition fields in the ROM's address: the `OP` field,
four bits wide, and the `uaddr` field, which is three bits wide. This field is
the microprogram counter used to address the currently executing
instruction. It is mandatory to include it, but its width is completely
arbitrary. `mcasm` will warn you if any of your microprograms are too big for
the chosen size of `uaddr`. The last `cond` defined is the least significant
part of the address.

The `signal` declarations define the output of the ROM. The bitfield for each
signal is specified in binary. You may use the traditional values 0 and 1. For
increased readability, you may also use the characters `.` or `-` for zero.

Conditionals are made up of letters, numbers, and non-whitespace characters
except colons (`:`), semicolons (`;`) and equal signs (`=`).

Signal names may contain any character except the equals sign (`=`). Notably,
they may include spaces (for legibility). A signal may not start with a minus
sign (`-`).

Signals starting with a slash (`/`) are recognised as active-low signals. The
corresponding bits are inverted when the ROM bitmap is written.

Signal fields may be named for convenience. Macros to extract field values will
be added to any C output generated (in future versions, more functionality will
be attached to them). Field definitions may use `X`, `x` or `+` to specify a
bit that's included in the field, and `.`, `_`, or `-` to specify bits that are
excluded from the field.

Microprograms are declared with the `start` keyword, followed by a
comma-separated set of conditionals and their values in the format
`cond=value`. If a particular field is immaterial to a microprogram, a don't
care value may be specified using the standard electronics don't care symbol,
`X`.

After the `start` keyword and condition specification, the microprogram
follows. Each line of the microprogram defines a comma separated list of zero
or more signal names as defined previously. The order of signals in each line
does not matter. Lines are terminated with semicolons (so that, in practice,
each may span multiple actual lines in the source code).

A special keyword used in some cases is `hold`. When encountered, it indicates
that the set of signals active on the previous line should be active on the
line where `hold` appears. In this case, some signals may be deactivated by
listing them prefixed with a minus. For example, the `-MREAD` specification on
line 3 of the microprogram above turns off the `MREAD` signal.

Multiple microprograms may be specified with multiple `start` keywords. When
using don't-care values, start with the least specific (most `X` bits)
microprograms and move on to the most specific ones.

## Restrictions

`mcasm` can generate microcode for architectures where microcode jumps around a
lot, but its syntax is more suited to architectures where microcode is executed
mostly sequentially with occasional jumps. If your architecture doesn't even
have the notion of a microprogram, `mcasm` may not suit you.


debian/changelog
mcasm (1.4.3) unstable; urgency=low

  * Added FIELD_ defines for multi-bit field values to the C output.

 -- Alexios Chouchoulas <alexios@bedroomlan.org>  Sun, 21 Jun 2020 12:34:54 +0300

mcasm (1.4.2) unstable; urgency=low

  * Fixed an error reporting bug.

 -- Alexios Chouchoulas <alexios@bedroomlan.org>  Sun, 6 Oct 2019 20:31:20 +0300

mcasm (1.4.1) unstable; urgency=low

  * Fixed URLs.
  * Stepped versions because 1.4 had packaging issues.

 -- Alexios Chouchoulas <alexios@bedroomlan.org>  Sat, 5 Oct 2019 13:54:34 +0300

mcasm (1.4) unstable; urgency=low

  * Printing out preprocessed microcode now includes microcode comments.

 -- Alexios Chouchoulas <alexios@bedroomlan.org>  Sat, 5 Oct 2019 13:22:35 +0300

mcasm (1.3) unstable; urgency=low

  * Ported to Python 3.
  * Improved command line parsing and code structure.
  * Fixed a couple of minor bugs.
  * Published on github.

 -- Alexios Chouchoulas <alexios@bedroomlan.org>  Sun, 5 May 2019 13:22:31 +0300

mcasm (1.2) unstable; urgency=low

  * Added Verilog source and altera MIF file support.

 -- Alexios Chouchoulas <alexios@bedroomlan.org>  Wed, 19 Dec 2012 23:02:47 +0200

mcasm (1.1) unstable; urgency=low

  * Added geshi syntax highlighter.
  * Added emacs mcasm-mode major mode.

 -- Alexios Chouchoulas <alexios@bedroomlan.org>  Mon, 24 Oct 2011 23:15:44 +0300

mcasm (1.0) unstable; urgency=low

  * Initial revision imported from CFT project.

 -- Alexios Chouchoulas <alexios@bedroomlan.org>  Mon, 24 Oct 2011 19:00:40 +0300


contrib/emacs/mcasm-mode.el
 ;;; mcasm.el --- Mcasm major mode

 ;; Copyright (C) 2001  Free Software Foundation, Inc.

 ;; Author: StefanMonnier
 ;; Keywords: extensions

 ;; This file is free software; you can redistribute it and/or modify
 ;; it under the terms of the GNU General Public License as published by
 ;; the Free Software Foundation; either version 2, or (at your option)
 ;; any later version.

 ;; This file is distributed in the hope that it will be useful,
 ;; but WITHOUT ANY WARRANTY; without even the implied warranty of
 ;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 ;; GNU General Public License for more details.

 ;; You should have received a copy of the GNU General Public License
 ;; along with GNU Emacs; see the file COPYING.  If not, write to
 ;; the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 ;; Boston, MA 02111-1307, USA.

 ;;; Commentary:

 ;; 

 ;;; Code:

 (defvar mcasm-mode-map
   (let ((map (make-sparse-keymap)))
     (define-key map [foo] 'mcasm-do-foo)
     map)
   "Keymap for `mcasm-mode'.")


 (defvar mcasm-mode-syntax-table
   (let ((st (make-syntax-table)))
     ;(modify-syntax-entry ?# "<" st)
     ;(modify-syntax-entry ?\n ">" st)
     (modify-syntax-entry ?_ "w" st)
     (modify-syntax-entry ?/ "w" st)

     (modify-syntax-entry ?\/ ". 12b" st)
     (modify-syntax-entry ?\n "> b" st)


     st)
   "Syntax table for `mcasm-mode'.")

 (defvar mcasm-font-lock-keywords
   '(("\\(start .+\\)" (1 font-lock-function-name-face))
     ("\\<\\(cond\\|field\\|hold\\|s\\(?:ignal\\|tart\\)\\|uaddr\\)\\>" (1 font-lock-keyword-face))
     ("\\(#\\sw+\\)" (1 font-lock-preprocessor-face))
     ("#\\sw+  *\\(\\sw+\\)" (1 font-lock-variable-name-face))
     ("field *\\(\\sw+\\)" (1 font-lock-constant-face))
     ("cond *\\(\\sw+\\)" (1 font-lock-variable-name-face))
     ("signal *\\([^ #\\,;-]+\\)" (1 font-lock-constant-face))
     )
   "Keyword highlighting specification for `mcasm-mode'.")

 ;;;###autoload
 (define-derived-mode mcasm-mode fundamental-mode "mcasm"
   "A major mode for editing mcasm files."
   :syntax-table mcasm-mode-syntax-table
   (set (make-local-variable 'comment-start) "// ")
   (set (make-local-variable 'comment-start-skip) "//+\\s-*")
   (set (make-local-variable 'font-lock-defaults)
	'(mcasm-font-lock-keywords))
   (set (make-local-variable 'indent-line-function) 'mcasm-indent-line)
   (set (make-local-variable 'imenu-generic-expression)
	mcasm-imenu-generic-expression)
   (set (make-local-variable 'outline-regexp) mcasm-outline-regexp)
   ;...
   )

 ;;; Indentation

 (defun mcasm-indent-line ()
   "Indent current line of Mcasm code."
   (interactive)
   (let ((savep (> (current-column) (current-indentation)))
	 (indent (condition-case nil (max (mcasm-calculate-indentation) 0)
		   (error 0))))
     (if savep
	 (save-excursion (indent-line-to indent))
       (indent-line-to indent))))

; (defun mcasm-calculate-indentation ()
;   "Return the column to which the current line should be indented."
;   ...)


 (provide 'mcasm-mode)

 ;;; mcasm.el ends here


;;; End of file.


contrib/geshi/mcasm.php
<?php
/*************************************************************************************
 * mcasm.php
 * -----
 * Author: Alexios Chouchoulas (alexios@bedroomlan.org)
 * 
 * Derived from c.php:
 *
 * Author: Nigel McNie (nigel@geshi.org)
 * Contributors:
 *  - Jack Lloyd (lloyd@randombit.net)
 * Copyright: (c) 2004 Nigel McNie (http://qbnz.com/highlighter/)
 * Release Version: 1.0.7.20
 * Date Started: 2004/06/04
 *
 * C language file for GeSHi.
 *
 * CHANGES
 * -------
 * 2004/XX/XX (1.0.4)
 *   -  Added a couple of new keywords (Jack Lloyd)
 * 2004/11/27 (1.0.3)
 *   -  Added support for multiple object splitters
 * 2004/10/27 (1.0.2)
 *   -  Added support for URLs
 * 2004/08/05 (1.0.1)
 *   -  Added support for symbols
 * 2004/07/14 (1.0.0)
 *   -  First Release
 *
 * TODO (updated 2004/11/27)
 * -------------------------
 *  -  Get a list of inbuilt functions to add (and explore C more
 *     to complete this rather bare language file
 *
 *************************************************************************************
 *
 *     This file is part of GeSHi.
 *
 *   GeSHi is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   GeSHi is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with GeSHi; if not, write to the Free Software
 *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 ************************************************************************************/

$language_data = array (
	'LANG_NAME' => 'mcasm',
	'COMMENT_SINGLE' => array(1 => '//'),
	'COMMENT_MULTI' => array('/*' => '*/'),
	'CASE_KEYWORDS' => GESHI_CAPS_NO_CHANGE,
	'QUOTEMARKS' => array("'", '"'),
	'ESCAPE_CHAR' => '\\',
	'KEYWORDS' => array(
		2 => array(
			   'cond', 'field', 'signal', 'start', 'uaddr'
			   ),
		3 => array(
			   '#define', '#ifdef', '#if', '#ifndef', '#endif'
			   ),
			    ),
	'SYMBOLS' => array(
			   '(', ')', '=', '-',
			   ),
	'CASE_SENSITIVE' => array(
				  GESHI_COMMENTS => true,
				  1 => false,
				  2 => false,
				  3 => false,
				  4 => false,
				  ),
	'STYLES' => array(
			  'KEYWORDS' => array(
					      1 => 'color: #b1b100;',
					      2 => 'color: #000000; font-weight: bold;',
					      3 => 'color: #000066;',
					      4 => 'color: #993333;'
					      ),
			  'COMMENTS' => array(
					      1 => 'color: #808080; font-style: italic;',
					      2 => 'color: #339933;',
					      'MULTI' => 'color: #808080; font-style: italic;'
					      ),
			  'ESCAPE_CHAR' => array(
						 0 => 'color: #000099; font-weight: bold;'
						 ),
			  'BRACKETS' => array(
					      0 => 'color: #66cc66;'
					      ),
			  'STRINGS' => array(
					     0 => 'color: #ff0000;'
					     ),
			  'NUMBERS' => array(
					     0 => 'color: #cc66cc;'
					     ),
			  'METHODS' => array(
					     1 => 'color: #202020;',
					     2 => 'color: #202020;'
					     ),
			  'SYMBOLS' => array(
					     0 => 'color: #66cc66;'
					     ),
			  'REGEXPS' => array(
					     0 => 'color: #007800;',
					     1 => 'color: #780000;',
					     ),
			  'SCRIPT' => array(
					    )
			  ),
	'URLS' => array(
			1 => '',
			2 => '',
			3 => '',
			4 => ''
			),
	'OOLANG' => false,
	'OBJECT_SPLITTERS' => array(
				    ),
	'REGEXPS' =>
	array(
	      0 => array(
			 GESHI_SEARCH => '(\S+)(\s*=)',
			 GESHI_REPLACE => '\\1',
			 GESHI_MODIFIERS => '',
			 GESHI_BEFORE => '',
			 GESHI_AFTER => '\\2',
			 ),
	      ),
	1 => array(
		   GESHI_SEARCH => '^(\s*cond\s*)(\S+)(\s*:)',
		   GESHI_REPLACE => '\\2',
		   GESHI_MODIFIERS => '',
		   GESHI_BEFORE => '\\1',
		   GESHI_AFTER => '\\3',
		   ),
	'STRICT_MODE_APPLIES' => GESHI_NEVER,
	'SCRIPT_DELIMITERS' => array(
				     ),
	'HIGHLIGHT_STRICT_BLOCK' => array(
					  ),
	'TAB_WIDTH' => 4
);

?>
